bitbots_vision/bitbots_vision/vision_modules/yoeo_handler.py

```diff
@@ -0,0 +1,207 @@
+from rclpy import logging
+from typing import List
+
+logger = logging.get_logger('vision_yoeo')
+
+try:
+    from yoeo import models as torch_models, detect as torch_detect
+    from yoeo.utils.utils import load_classes
+except ImportError:
+    logger.error("Not able to import pytorchyolo. This might be fine if you use another method.")
+
+
+class YOEOHandlerTemplate(ABC):
+    def __init__(self, config: dict, model_path: str):
+        self._use_caching = config['caching']
+
+        self._image = None
+        self._iou_non_max_suppression_thresh = config['yoeo_nms_threshold']
+
+        self._det_candidates = defaultdict(list)
+        self._det_class_names = None
+        self._det_confidence_thresh = config['yoeo_conf_threshold']
+
+        self._segmentation = None
+        self._segmentations = dict()
+        self._seg_class_idx = dict()
+
+        self._load_candidate_class_names(model_path)
+
+    def _load_candidate_class_names(self, model_path: str) -> None:
+        path = join(model_path, "data", "yoeo_names.yaml")
+        data = load_classes(path)
+
+        self._det_class_names = data['detection']
+        for key, i in zip(data['segmentation'], range(len(data['segmentation']))):
+            self._seg_class_idx[key] = i
+
+    def get_candidates(self, class_name: str) -> List[Candidate]:
+        """
+        To comply with YoloHandler interface
+        """
+        return self.get_detection_candidates_for(class_name)
+
+    def get_detection_candidates_for(self, class_name: str) -> List[Candidate]:
+        assert class_name in self._det_class_names, \
+            f"Class '{class_name}' is not available for the current YOEO model (detection)"
+
+        self.predict()
+
+        return self._det_candidates[class_name]
+
+    def predict(self) -> None:
+        if self._prediction_has_to_be_updated():
+            self._compute_new_prediction()
+
+    def _prediction_has_to_be_updated(self) -> bool:
+        return self._det_candidates is None or self._segmentation is None or not self._use_caching
+
+    def get_classes(self) -> List[str]:
+        """
+        To comply with YoloHandler interface
+        """
+        return self.get_detection_candidate_class_names()
+
+    def get_detection_candidate_class_names(self) -> List[str]:
+        return self._det_class_names
+
+    def get_segmentation_candidate_class_names(self):
+        return self._seg_class_idx
+
+    def get_segmentation_for(self, class_name: str):
+        assert class_name in self._seg_class_idx, \
+            f"Class '{class_name}' ist not available for the current YOEO model (segmentation)"
+
+        self.predict()
+
+        if class_name not in self._segmentations.keys():
+            self._compute_segmentation_for(class_name)
+
+        return self._segmentations[class_name]
+
+    def _compute_segmentation_for(self, class_name: str) -> None:
+        segmentation = where(self._segmentation == self._seg_class_idx[class_name], 1, 0)
```

Comment:
The query would look like this `self._seg_class_idx.index(class_name)` if a list instead of a dict is used.